### PR TITLE
logging: Add option for prolonged backend initialization

### LINKED ISSUE #38494: Flooded logs when using CDC_ACM as back-end
**Is your enhancement proposal related to a problem? Please describe.**
If USB CDC_ACM is selected as logging back-end and the USB cable is not connected to the board at the start up, logs could be flooded with `LOG_WRN("Device not configured or suspended, drop %d bytes", len);`.

**Describe the solution you'd like**
The best option would be for logger back-end not to try to send anything on CDC_ACM back-end, until its fully operational and ready to send the data (backend_ready_callback).

**Describe alternatives you've considered**
* Changing logging priority of the CDC_ACM class so that no warning logs would be processed.
* Delaying logger processing thread with `CONFIG_LOG_PROCESS_THREAD_STARTUP_DELAY_MS`

Both above alternatives are workaround and are not satisfactory solutions.

**Additional context**
AFAIK CDC_ACM is using uart API so from logger perspective anything could be connected there and there is no good mechanism to know which back-end is chosen (except for Kconfig option).
To get above enhancement in i would propose to, at least, add waiting for DTR in logger subsystem before trying to put any character on usb line.

Alternatively (more complex but i think more accurate):
1. Separate setting usb_status_callback from usb_enable(). Add usb_register_status_callback() or whatsoever similar. 
2. Allow to set multiple status callbacks for single USB device.
3. In case USB CDC_ACM is used as logging back-end, use newly introduced function to register back-end callback during logger initialization.
4. Add USB state tracking in logger subsystem and allow to send the data only if the back-end is ready.

This could be more generic for other back-ends where the device may not always be ready to send the data.
I think the assumption that the back-end is always ready is incorrect.

### PR DESCRIPTION
**Pull request by nordic-krch**
Extended logging backend API with log_backend_is_ready call which
returns 0 is backend is ready. Logging core will make sure that
all autostarted backends are ready before they are enabled.
This option allows to handle backends which are not yet ready
after init function is called (e.g. usb backend that is not plugged
in). If this is the only backend in the system, logging processing
will not start util first backend is ready.
    
Function for checking readiness is optional and when backend has
no such function it is assumed that backend is ready after
initialization function returns which makes this feature backward
compatible.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

PR sets instruments that could be used to modify uart backend and fix https://github.com/zephyrproject-rtos/zephyr/issues/38494

### CODE REVIEW

**Comment by nordic-krch** [comment]
@emob-nordic fyi this is the first step. Next will be to add support for dtr line check in uart backend.

File: include/logging/log_backend.h [hunk 1]
```diff
@@ -95,6 +96,45 @@ extern const struct log_backend __log_backends_end[];
...
+ */
+static inline bool log_backend_is_ready(const struct log_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+	if (backend->api->is_ready) {
+		return backend->api->is_ready(backend);
+	}
+
+	return 0;
+}
```
Comment (dcpleung) [inline include/logging/log_backend.h:136]: Since this is boolean, shouldn't returning `true` when backend is ready be conceptually easier to understand?

Comment (tejlmand) [reply include/logging/log_backend.h:136]: from the description of the function it doesn't sound like it should be bool
```
 * @param[in] backend  Pointer to the backend instance.
 *
 * @retval 0 if backend is ready.
 * @retval -EBUSY if backend is not yet ready.
```

**Comment by carlescufi** [comment]
@nordic-krch ping

**Pull request closed without merging**